src/Servers.jl

```diff
@@ -410,104 +307,79 @@ function handle_connection(f::Function, c::Connection;
     return
 end
 
+"creates an async thread that waits a specified amount of time before closing the connection"
+function check_readtimeout(c, readtimeout, wait_for_timeout)
+    @async while wait_for_timeout[]
+        if inactiveseconds(c) > readtimeout
+            @warn "Connection Timeout: $c"
+            try
+                writeheaders(c.io, Response(408, ["Connection" => "close"]))
+            finally
+                close(c)
+            end
+            break
+        end
+        sleep(8 + rand() * 4)
+    end
+    return
+end
 
 """
-Create a `HTTP.Stream` and parse the Request headers from a `HTTP.Transaction`.
-If there is a parse error, send an error Response.
-Otherwise, execute stream processing function `f`.
+Transaction handler: creates a new Stream for the Transaction, calls startread on it,
+then dispatches the stream to the user-provided handler function. Catches errors on all
+IO operations and closes gracefully if encountered.
 """
-function handle_transaction(f::Function, t::Transaction;
-                            final_transaction::Bool=false,
-                            verbose::Bool=false, kw...)
-
-    request = HTTP.Request()
-    http = Streams.Stream(request, t)
+function handle(h, t::Transaction, last::Bool=false)
+    request = Request()
+    http = Stream(request, t)
 
     try
         startread(http)
     catch e
-        # @show typeof(e)
-        # @show fieldnames(e)
         if e isa EOFError && isempty(request.method)
             return
-# FIXME https://github.com/JuliaWeb/HTTP.jl/pull/178#pullrequestreview-92547066
-#        elseif !isopen(http)
-#            @warn "Connection closed"
-#            return
-        elseif e isa HTTP.ParseError
-            @error e
+        elseif e isa ParseError
             status = e.code == :HEADER_SIZE_EXCEEDS_LIMIT  ? 413 : 400
             write(t, Response(status, body = string(e.code)))
             close(t)
             return
+        elseif e isa Base.IOError && e.code == -54
+            # read: connection reset by peer (ECONNRESET)
+            return
```

Comment:
So the client disappeared while we were still reading the headers?
I guess silent nothing is not a terrible thing to do.

I wonder if the log message on line 264 should say "ECONNRESET" in this case so that you can see in the server logs that you have a misbehaving client. We are still in the connection level task here right? (the transaction task hasn't started yet) so we could catch this up near line 264 and log it there too?

Reply:
Yes, I think people will run into this because in kubernetes, the default "health check" is to just open a tcp connection on a specified port (but not perform an actual http request). I'll move the catching/logging up as suggested.